src/middlewared/middlewared/plugins/pool_/import_pool.py

```diff
@@ -204,117 +204,172 @@ async def import_pool(self, job, data):
+                self.logger.warning('FAILED unhandled exception creating %r', ZPOOL_CACHE_FILE, exc_info=True)
+                set_cachefile_property = False
+
+        # We need to do as little zfs I/O as possible since this method
+        # is being called by a systemd service at boot-up. First step of
+        # doing this is to simply try to import all zpools that are in our
+        # database. Handle each error accordingly instead of trying to be
+        # fancy and determine which ones are "offline" since...in theory...
+        # all zpools should be offline at this point.
+        for i in self.middleware.call_sync('datastore.query', 'storage.volume'):
+            name, guid = i['vol_name'], i['vol_guid']
+            if not self.import_on_boot_impl(name, guid, set_cachefile_property):
+                continue
 
-            unlock_job = self.middleware.call_sync(
-                'pool.dataset.unlock', pool['name'], {'recursive': True, 'toggle_attachments': False}
-            )
-            unlock_job.wait_sync()
-            if unlock_job.error or unlock_job.result['failed']:
-                failed = ', '.join(unlock_job.result['failed']) if not unlock_job.error else ''
-                self.logger.error(
-                    f'Unlocking encrypted datasets failed for {pool["name"]} pool'
-                    f'{f": {unlock_job.error}" if unlock_job.error else f" with following datasets {failed}"}'
-                )
+            self.unlock_on_boot_impl(name)
 
-            # Child unencrypted datasets of root dataset would be mounted if root dataset is still locked,
-            # we don't want that
-            if self.middleware.call_sync(
-                'pool.dataset.get_instance_quick', pool['name'], {'encryption': True}
-            )['locked']:
-                with contextlib.suppress(CallError):
-                    self.middleware.call_sync('zfs.dataset.umount', pool['name'], {'force': True})
-
-                pool_mount = os.path.join('/mnt', pool['name'])
-                if os.path.exists(pool_mount):
-                    # We would like to ensure the path of root dataset has immutable flag set if it's not locked
-                    try:
-                        self.middleware.call_sync('filesystem.set_immutable', True, pool_mount)
-                    except CallError as e:
-                        self.logger.error('Unable to set immutable flag at %r: %s', pool_mount, e)
-
-        with contextlib.suppress(OSError):
-            os.unlink(ZPOOL_KILLCACHE)
-
-        if os.path.exists(ZPOOL_CACHE_FILE):
-            shutil.copy(ZPOOL_CACHE_FILE, zpool_cache_saved)
-
-        # Now finally configure swap to manage any disks which might have been removed
-        self.middleware.call_sync('disk.swaps_configure')
-        self.middleware.call_hook_sync('pool.post_import', None)
-        job.set_progress(100, 'Pools import completed')
+        # no reason to wait on these to complete before returning so call them
```

Comment:
There are reasons. For example, we generate a huge part of `etc` files in this hook. If we allow the boot process to continue, we'll be prone to a number of race conditions with services starting with invalid configuration.

Reply:
@themylogin I'm not changing anything with the order of this hook. I'm simply not blocking the entire boot-up process waiting for it to complete. The huge part of `etc` is still generated _after_ the zpools have been fully imported

Reply:
@yocalebo but we have to wait for this hook to complete before continuing with the boot process, otherwise we leave a bunch of files in /etc not generated

Reply:
hence my comment in the body of this PR about significant design problems. This is boot up, so we need to do as little as possible so we don't unnecessarily halt boot. Trying to do _everything_ at boot is really, really hurting us. I'll move that call_hook back out of the background but this is something that we need to fix. There is no reason to halt boot-up just so we can generate all these configurations.